torch/csrc/autograd/TraceTypeManual.cpp

```diff
@@ -0,0 +1,184 @@
+#include <ATen/TracerMode.h>
+#include <ATen/core/op_registration/op_registration.h>
+#include <c10/core/ScalarType.h>
+#include <c10/util/Optional.h>
+#include <torch/csrc/jit/frontend/tracer.h>
+#include <torch/csrc/utils/memory.h>
+#include <torch/library.h>
+
+using namespace at;
+
+namespace torch { namespace TraceType {
+
+namespace {
+
+Tensor & copy_(Tensor & self, const Tensor & src, bool non_blocking) {
+  jit::Value* output = nullptr;
+#if !defined(PYTORCH_DISABLE_TRACING)
+  if(torch::jit::tracer::isTracing()) {
+    const jit::tracer::TracingState& state = *jit::tracer::getTracingState();
+    auto& graph = state.graph;
+    if (state.force_outplace && self.storage().use_count() <= 1) {
+      // if you have no views of self, then an in place copy is equivalent to
+      // making sure we expand src to the same size as self
+      jit::Node* node = graph->create(jit::aten::expand_as, /*num_outputs=*/1);
+      jit::tracer::addInputs(node, "src", src);
+      jit::tracer::addInputs(node, "self", self);
+      graph->insertNode(node);
+      output = node->output();
+    } else {
+      output = graph->insert(
+          jit::aten::copy_,
+          {jit::tracer::getValueTrace(self), jit::tracer::getValueTrace(src)});
+      jit::tracer::recordSourceLocation(output->node());
+    }
+    jit::tracer::ensureUniqueIfOutOfPlaced("copy_ (possibly due to an assignment)", self);
+  }
+#endif
+
+  static auto op = c10::Dispatcher::singleton()
+      .findSchemaOrThrow("aten::copy_", "")
+      .typed<Tensor & (Tensor &, const Tensor &, bool)>();
+  {
+    at::tracer::impl::NoTracerDispatchMode tracer_guard;
```

Comment:
Out of curiosity, what is the reason we re-dispatch is done this way and not the same way it is done for VariableType which is:
```
  {
    at::AutoNonVariableTypeMode non_var_type_mode(true);
    self_.copy_(src_, non_blocking);
  }
```

Reply:
Short answer: yes, I probably should change to the more concise version here :)

More context:

There are 3 different ways to toggle tracing:
  1) c10 redispatch API. It dispatches to the lower level kernel (e.g. Tracer -> Autograd) by masking off the current dispatch key (Tracer). The mask only applies to this single call to avoid infinitely looping back to the Tracer kernel. It doesn't affect subsequent dispatches within the dispatch, i.e. op A Tracer --> op A CPU --> op B Tracer (it doesn't prevent it from dispatching to B's Tracer kernel).

  2) `NoTracerDispatchMode`. Similar to 1), except that it disables Tracer key for all subsequent dispatches within the scope.

  3) `jit::tracer::setTracingState(nullptr);` This is a domain switch to stop tracing.

We use different combinations of these switches at different places, which I plan to clean up next:

  - For generated TracerType kernels, we use 1) + 3) [Example](https://codebrowser.bddppq.com/pytorch/pytorch/torch/csrc/autograd/generated/TraceType_0.cpp.html#34). 1) is a "domain-neural" way to do redispatching as in some cases we do want to call the same level kernel for other ops being called by the op, e.g. Profiler. The domain specific decision that we should skip all Tracing for intermediate ops is achieved by 3).

  - At some other places I used the `NoTracerDispatchMode` guard, usually it's where `AutoNonVariableTypeMode` was originally used. More details: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/TracerMode.h#L72

  - One difference between these manually registered kernels and generated kernels is that the original VariableTypeManual uses `AutoNonVariableTypeMode` instead of `setTracingState(null)` - `NoTracerDispatchMode` guard is inherited from there.

I guess I should choose 2) for this case for it's simplicity while thinking about what's the desired consistency.

Reply:
Thanks for the detailed explanation!
So there is a crucial difference: The masking only effect the current dispatch but not the subsequent ones when using 1.
In the VariableType case, we actually never want this behavior: we either handle the autograd (and so the child ops should never do it) or not (and the child ops should always do it).

So for the tracer it is more complex as you might want to nest the tracing. Sounds good!

Reply:
@albanD yeah, it is exactly the key difference between redispatch v.s. guard.

For tracing specifically, it was originally inside VariableType so it automatically inherited no-nest-tracing behavior from the `AutoNonVariableTypeMode` guard. Besides, it also explicitly sets `setTracingState(nullptr)` before redispatching (except for these manually registered kernels which I'm not sure it's intentional or not). These two were overlapped but were added for different reasons.

So no-nest-tracing might actually be the invariant we want to stick with - as current tracing is for repeating the same behavior and tracing root ops are sufficient for this purpose. I simply haven't decided which pattern it should converge to. Now we have:
  * 1) + 3) for generated kernels;
  * 1) + 2) for these manually registered kernels; but 2) is totally a super set of 1) so I decided to use 2) only here as you questioned earlier;

Maybe changing everything to 2) is the right final state to move towards?

@ezyang @bhosmer - we talked about having a re-dispatching API that masks the dispatch-key for both current and subsequent calls, which is another option to converge to (if our goal is to reduce # of guards in the codebase :) )